qiskit/transpiler/passes/basis/basis_translator.py

```diff
@@ -210,100 +197,152 @@ def _basis_search(equiv_lib, source_basis, target_basis, heuristic):
+                # if we start from source gates and apply `basis_transforms` in reverse order, we'll end
+                # up with gates in the target basis. Note though that `basis_transforms` may include
+                # additional transformations that are not required to map our source gates to the given
+                # target basis.
+                self._basis_transforms.reverse()
+                raise StopIfBasisRewritable
+
+        def examine_edge(self, edge):
+            _, target, edata = edge
+            if edata is None:
+                return
+
+            index = edata["index"]
+            self._num_gates_remain_for_rule[index] -= 1
+
+            target = self.graph[target]
+            # if there are gates in this `rule` that we have not yet generated, we can't apply
+            # this `rule`. if `target` is already in basis, it's not beneficial to use this rule.
+            if self._num_gates_remain_for_rule[index] > 0 or target in self.target_basis:
+                raise retworkx.visit.PruneSearch
+
+        def edge_relaxed(self, edge):
+            _, target, edata = edge
+            if edata is not None:
+                gate = self.graph[target]
+                self._predecessors[gate] = edata["rule"]
+
+        def edge_cost(self, edge):
+            """Returns the cost of an edge.
+
+            This function computes the cost of this edge rule by summing
+            the costs of all gates in the rule equivalence circuit. In the
+            end, we need to subtract the cost of the source since `dijkstra`
+            will later add it.
+            """
+
+            if edge is None:
+                # the target of the edge is a gate in the target basis,
+                # so we return a default value of 1.
+                return 1
+
+            cost_tot = 0
+            rule = edge["rule"]
+            for gate, qargs, _ in rule.circuit:
+                key = Key(name=gate.name, num_qubits=len(qargs))
+                cost_tot += self._opt_cost_map[key]
+
+            source = edge["source"]
+            return cost_tot - self._opt_cost_map[source]
+
+        @property
+        def basis_transforms(self):
+            """Returns the gate basis transforms."""
+            return self._basis_transforms
+
+    all_gates_in_lib = set()
+
+    graph = retworkx.PyDiGraph()
+    nodes_to_indices = dict()
+    num_gates_for_rule = dict()
+
+    def get_nid_or_insert(key):
+        if key not in nodes_to_indices:
+            nodes_to_indices[key] = graph.add_node(key)
+        return nodes_to_indices[key]
+
+    rcounter = 0  # running sum of the number of equivalence rules in the library.
+    for key in equiv_lib._get_all_keys():
+        target = get_nid_or_insert(key)
+        all_gates_in_lib.add(key)
+        for equiv in equiv_lib._get_equivalences(key):
+            sources = set(
+                Key(name=gate.name, num_qubits=len(qargs)) for gate, qargs, _ in equiv.circuit
+            )
```

Comment:
```suggestion
            sources = {
                Key(name=gate.name, num_qubits=len(qargs)) for gate, qargs, _ in equiv.circuit
            }
```
set comprehensions are non-trivially more efficient than function calls over generator expressions for some reason.